trace/stats.go

```diff
@@ -0,0 +1,172 @@
+package trace
+
+import (
+	"sync"
+
+	"github.com/akitasoftware/akita-cli/pcap"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+)
+
+// We produce a set of packet counters indexed by interface and
+// port number (*either* source or destination.)
+type PacketCounters struct {
+	// Flow
+	Interface string
+	SrcPort   int
+	DstPort   int
+
+	// Number of events
+	TCPPackets    int
+	HTTPRequests  int
+	HTTPResponses int
+	Unparsed      int
+}
+
+func (c *PacketCounters) Add(d PacketCounters) {
+	c.TCPPackets += d.TCPPackets
+	c.HTTPRequests += d.HTTPRequests
+	c.HTTPResponses += d.HTTPResponses
+	c.Unparsed += d.Unparsed
+}
+
+// A consumer accepts incremental updates in the form
+// of PacketCounters.
+type PacketCountConsumer interface {
+	// Add an additional measurement to the current count
+	Update(delta PacketCounters)
+
+	// Signal the end of the measurement interval
+	Close()
+}
+
+// Discard the count
+type PacketCountDiscard struct {
+}
+
+func (d *PacketCountDiscard) Update(_ PacketCounters) {
+}
+
+func (d *PacketCountDiscard) Close() {
+}
+
+// A consumer that sums the count by (interface, port) pairs.
+// In the future, this could put counters on a pipe and do the increments
+// in a separate goroutine, but we would *still* need a mutex to read the
+// totals out.
+// TODO: limit maximum size
+type PacketCountSummary struct {
+	total       PacketCounters
+	byPort      map[int]*PacketCounters
+	byInterface map[string]*PacketCounters
+	mutex       sync.RWMutex
+}
+
+func NewPacketCountSummary() *PacketCountSummary {
+	return &PacketCountSummary{
+		byPort:      make(map[int]*PacketCounters),
+		byInterface: make(map[string]*PacketCounters),
+	}
+}
+
+func (s *PacketCountSummary) Update(c PacketCounters) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if prev, ok := s.byPort[c.SrcPort]; ok {
+		prev.Add(c)
+	} else {
+		new := &PacketCounters{
+			Interface: "*",
+			SrcPort:   c.SrcPort,
+			DstPort:   0,
+		}
+		new.Add(c)
+		s.byPort[new.SrcPort] = new
+	}
+
+	if prev, ok := s.byPort[c.DstPort]; ok {
+		prev.Add(c)
+	} else {
+		// Use SrcPort as the identifier in the
+		// accumualted counter
```

Comment:
Is the SrcPort/DstPort distinction used anywhere?

Also, typo: accumulated.

Reply:
The original data has SrcPort and DstPort filled in.

The problem is that I couldn't think of a good way to summarize all the port-pairs compactly, so this implementation doesn't track them.  (It might be that the right thing is to collect the entire 4-tuple so we can track packets per flow?)  But I think they should be filled in correctly at the source of data in case we figure out how to do better.